### DISCUSSION TITLE
How to apply fallback font to page.insert_text

### DISCUSSION THREAD

**Discussion by smithah** [pymupdf/PyMuPDF#3659] [Looking for help]
There is an issue where Noto Sans Devanagari Regular is printing boxes for English Characters, with page.insert_text, is there a way to add fallback font for insert_text ? Please let me know

**Comment by JorjMcKie** [comment]
No, there is no way to do that for `insert_text`/`insert_textbox`. `TRextWriter.fill_textbox()` support this.

Use `Page.insert_htmlbox` instead. This is BTW the only method to benefit from "text shaping" as supplied by the Harfbuzz library. Especially for text using Devanagari I cannot imagine how you are getting along with a character-based insertion method.

**Reply by smithah** [reply]
Thanks, character based insertion is working well with insert_text and it's faster. Using Noto Sans Devanagari,. It's filling up the white space in this method. Except English Characters are printed as boxes.

As per Discussion #3634 

Tried quite a few approaches to reduce the spacing in the boxes. But the co-ordinates for rectangles are coming in from a predetermined source where placement of text to be done on blank  pdf with some images on it already.

With insert_htmlbox Indic languages are broken character wise. Here insert_text 
works very well except as mentioned above English Characters are printed as boxes. Hence looking to handle those with fallback font.

**Reply by JorjMcKie** [reply]
Ok, it is your decision in the end.
Again my recommendation **_not_** to using Noto Sans Devanagari but FiraGo (contained in pymupdf-fonts). It supports both Devanagari and Extended Latin.

**Reply by smithah** [reply]
Thanks will use that and update here.
